sql/src/test/java/ddlog/RuntimeTest.java

```diff
@@ -0,0 +1,1064 @@
+        testTranslation(query, program, true);
+    }
+
+    @Test
+    public void testScope() {
+        String query = "create view v1 as SELECT DISTINCT t1.column1 FROM t1";
+        String program = this.header(false) +
+                "typedef Ttmp0 = Ttmp0{column1:signed<64>}\n" +
+                this.relations(false) +
+                "relation Rtmp0[Ttmp0]\n" +
+                "output relation Rv1[Ttmp0]\n" +
+                "Rv1[v2] :- Rt1[v0],var v1 = Ttmp0{.column1 = v0.column1},var v2 = v1.";
+        testTranslation(query, program);
+    }
+
+    @Test
+    public void testScopeWNulls() {
+        String query = "create view v1 as SELECT DISTINCT t1.column1 FROM t1";
+        String program = this.header(true) +
+            "typedef Ttmp0 = Ttmp0{column1:Option<signed<64>>}\n" +
+            this.relations(true) +
+            "relation Rtmp0[Ttmp0]\n" +
+            "output relation Rv1[Ttmp0]\n" +
+            "Rv1[v2] :- Rt1[v0],var v1 = Ttmp0{.column1 = v0.column1},var v2 = v1.";
+        testTranslation(query, program, true);
+    }
+
+    @Test
+    public void testImplicitJoin() {
+        String query = "create view v1 as SELECT DISTINCT * FROM t1, t2";
+        String program = this.header(false) +
+                "typedef Ttmp0 = Ttmp0{column1:signed<64>, column2:string, column3:bool, column10:signed<64>}\n" +
+                this.relations(false) +
+                "output relation Rv1[Ttmp0]\n" +
+                "Rv1[v3] :- Rt1[v0],Rt2[v1],true,var v2 = Ttmp0{.column1 = v0.column1,.column2 = v0.column2,.column3 = v0.column3,.column10 = v1.column1},var v3 = v2.";
+        testTranslation(query, program);
+    }
+
+    @Test
+    public void testJoinStar() {
+        String query = "create view v1 as SELECT DISTINCT * FROM t1 JOIN t2 ON t1.column1 = t2.column1";
+        String program = this.header(false) +
+                "typedef Ttmp0 = Ttmp0{column1:signed<64>, column2:string, column3:bool, column10:signed<64>}\n" +
+                this.relations(false) +
+                "output relation Rv1[Ttmp0]\n" +
+                "Rv1[v3] :- Rt1[v0],Rt2[v1],(v0.column1 == v1.column1),true,var v2 = Ttmp0{.column1 = v0.column1,.column2 = v0.column2,.column3 = v0.column3,.column10 = v1.column1},var v3 = v2.";
+        testTranslation(query, program);
+    }
+
+    @Test
+    public void testJoinStarWNull() {
+        String query = "create view v1 as SELECT DISTINCT * FROM t1 JOIN t2 ON t1.column1 = t2.column1";
+        String program = this.header(true) +
+                "typedef Ttmp0 = Ttmp0{column1:Option<signed<64>>, column2:Option<string>, column3:Option<bool>, column10:Option<signed<64>>}\n" +
+                this.relations(true) +
+                "output relation Rv1[Ttmp0]\n" +
+                "Rv1[v3] :- Rt1[v0],Rt2[v1],unwrapBool(a_eq_NN(v0.column1, v1.column1)),true,var v2 = Ttmp0{.column1 = v0.column1,.column2 = v0.column2,.column3 = v0.column3,.column10 = v1.column1},var v3 = v2.";
```

Comment:
I assume the ",true," and "val v2 = v3" clauses get optimized away by ddlog/rust?

Reply:
If they aren't they should. This is a legal program, even if it is not optimal.